magento/magento2
# Improve error messaging when merged XML is invalid

### Description
When dealing with XML in Magento, the error messages are not always most helpful. When each XML document is individually valid, but the result of merging these isn't, the error message does not help a developer identify the issue.

This pull request adds the filename causing a problem to the error message output.

<details open><summary>Before and after pictures</summary>

The following two pictures show the output from `bin/magento setup:upgrade` where the problem shows up.

Before:

![Screenshot_2022-10-04_15-59-12](https://user-images.githubusercontent.com/334786/193854121-559288ca-d6a8-4531-8809-94f929f5a928.png)

After:

![Screenshot_2022-10-04_15-57-42](https://user-images.githubusercontent.com/334786/193854777-528c8e33-7d98-4a6b-b360-10e946f06fd4.png)

</details>

### Manual testing scenarios
Note that if you fix/merge https://github.com/magento/magento2/pull/36240 before this, then you will need to find alternative steps to reproduce.
1. Ensure Magento is in "developer" mode with `bin/magento deploy:mode:set developer`
1. Create a new local module in `app/code/`
1. Create a valid `etc/db_schema.xml` file changing the 'type' of an existing column from 'int' (with 'padding') to 'varchar'. For example, https://github.com/magento/magento2/blob/447c613c86ea655d94f511b495fcc5d1273072c3/app/code/Magento/AsynchronousOperations/etc/db_schema.xml#L20-L21 and
   ```xml
   <?xml version="1.0"?>
   <schema xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
           xsi:noNamespaceSchemaLocation="urn:magento:framework:Setup/Declaration/Schema/etc/schema.xsd">
       <table name="magento_bulk">
           <column xsi:type="varchar" name="operation_count" />
       </table>
   </schema>
   ```
1. Run `bin/magento setup:upgrade`
1. Witness error message

Before this change, it is very difficult to identify the location of the error.

### Questions or comments
All unit tests pass both before and after this change, suggesting that the code coverage isn't ideal. Perhaps we need a separate task to improve code coverage of this class.

### Contribution checklist (*)
